core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala

```diff
@@ -87,15 +91,39 @@ abstract class ConfigReaderFailure {
 /**
  * A failure representing the inability to convert a null value. Since a null
  * represents a missing value, the location of this failure is always None.
+ *
+ * @param candidates a set of candidate keys that might map to the desired value
+ *                   in case of a misconfigured ProductHint
  */
-final case object CannotConvertNull extends ConfigReaderFailure {
+final case class CannotConvertNull(candidates: Set[String] = Set()) extends ConfigReaderFailure {
   val location = None
   val path = None
 
   def description = "Cannot convert a null value."
 
   def withImprovedContext(parentKey: String, parentLocation: Option[ConfigValueLocation]) =
-    KeyNotFound(parentKey, parentLocation)
+    KeyNotFound(parentKey, parentLocation, candidates)
+}
+
+object CannotConvertNull {
+  @tailrec
+  private[this] def isSubsequence(s1: String, s2: String): Boolean = {
+    if (s1.isEmpty())
+      true
+    else if (s2.isEmpty())
+      false
+    else if (s1.head == s2.head)
+      isSubsequence(s1.tail, s2.tail)
+    else
+      isSubsequence(s1, s2.tail)
+  }
+
+  def apply(fieldName: String, confObj: ConfigObject): CannotConvertNull = {
```

Comment:
Minor suggestion: consider receiving here an `Iterable[String]` or even a `TraversableOnce[String]` with all available keys instead of a `ConfigObject`. While we always have access to a `ConfigObject` in the places we use it, users might want to create their own `ConfigReader` instances and that way we wouldn't be so restrictive.